daft/execution/shuffles/pre_shuffle_merge.py

```diff
@@ -58,42 +61,78 @@ def pre_shuffle_merge(
         done_with_input = no_more_input and len(materialized_maps) == len(in_flight_maps)
 
         if enough_maps or done_with_input:
-            # Initialize the first merge group
-            merge_groups = []
-            current_group = [materialized_maps[0][0]]
-            current_size = materialized_maps[0][1]
-
-            # Group remaining maps based on memory threshold
-            for partition, size in materialized_maps[1:]:
-                if current_size + size > pre_shuffle_merge_threshold:
-                    merge_groups.append(current_group)
-                    current_group = [partition]
-                    current_size = size
-                else:
-                    current_group.append(partition)
-                    current_size += size
+            # Get location information for all materialized partitions
+            partitions = [m[0].result().partition() for m in materialized_maps]
+            location_map = ray.experimental.get_object_locations(partitions)
+
+            # Group partitions by node
+            node_groups = defaultdict(list)
+            unknown_location_group = []  # Special group for partitions without known location
 
-            # Add the last group if it exists and is either:
-            # 1. Contains more than 1 partition
-            # 2. Is the last group and we're done with input
-            # 3. The partition exceeds the memory threshold
-            if current_group:
-                if len(current_group) > 1 or done_with_input or current_size > pre_shuffle_merge_threshold:
-                    merge_groups.append(current_group)
+            for partition, size in materialized_maps:
+                partition_ref = partition.partition()
+                location_info = location_map.get(partition_ref, {})
+
+                if not location_info or "node_ids" not in location_info or not location_info["node_ids"]:
+                    unknown_location_group.append((partition, size))
+                else:
+                    node_id = location_info["node_ids"][0]  # Use first node if multiple locations exist
```

Comment:
how often does a partition exist on multiple nodes?

maybe we should have a strat like picking the node with the least memory usage or the one that is closest to a merge?

Reply:
Not too sure, tbh. In fact I can't think off the top of my head a situation where it would happen, some object would have to be shared between map partition tasks for that to happen.

I can add a todo for this though.